### ISSUE TITLE
Extremely long compile time: 24h and running for a 270 mb typ file

### ISSUE THREAD

**Issue by chrarnold** [typst/typst#4501]
### Description

my `typst compile` command on my Mac M2 runs for a full 24h already for a 270 mb file. Admittedly, it is a rather unusual document with one line only, consisting of around 57 million characters (many of which are formatted individually with underscore and bold weight, along with a few hundred footnotes and a few hundred blocks of 100-10000 characters with background colors), but is it normal? Is is better to split it up in multiple lines? Currently the one lines has a space every 185 characters, but it is still only one line.

I'd be interested to learn whether that's still "normal", or a bug, or a consequence of a design decision that in my case causes the very long compilation time.

Thanks for the input!

### Typst version

- [ ] I am using the latest version of Typst

**Comment by Enivex** [comment]
Is this on 0.11.1, or on main? There have been quite a few changes since the latest release that should affect performance.

That being said, I suspect this document would require a ridiculous amount of memory to compile (possibly on the order of 1 TB). If you max out your ram and it starts using the swap file, performance will slow to a crawl.

**Comment by chrarnold** [comment]
I am running on 0.11.1, yes! Good point about the memory - According to the Activity Monitor, typst uses currently around 23.5 GB RAM, I have a 64 GB RAM M2 here - and around 50 GB are currently used in total - so it looks like it is not swapping!?

What's the reason it may use so much memory, though? Can I speed it up by splitting lines? I cant change the 56 million base characters, and a significant fraction of those needs additional formatting. Currently, however, I am re-defining underline-formatting over and over again, as opposed to defining it once in the header section.  This adds a lot of extra characters in the typst file, but not sure it makes a difference for the compilation time. Currently, I am putting 2.5 million times a block like this, as opposed to defining it once for all elements:

`#underline(stroke: (thickness: 0.5pt, paint: red, cap: "square"))[#text(weight: "bold")[bla]]`

I'd like to understand whats the main bottleneck for the compilation time. From my empirical testing, the 57 million characters are compiled very fast (without the extra formatting), but adding the extra underline blocks make it very slow. Will it be faster when I define different "standards" (essentially reducing each block to `#underline[bla]` ), which would save some characters in the typst file?

**Comment by Dherse** [comment]
Speaking from experience:
- Check that you're not storing the result of a query in a state, or any other kind of introspection loop
- If you are, change it by always doing the query "in-line" so-to-speak
- You might not think that you are, but try disabling components that run any kind of query and see if it fixes it, do trial-and-error until you find the query causing it
- Since you're on an M2 Mac (the model doesn't matter much), it's likely not a computational limit

If you are not in the aforementioned case, then may I ask whether you are running anything weird? a package that is particularly heavy, etc.

Finally, if you can share the document, I would be open to helping you figure it out :) Or if you can share the `--timings` output (although for a 24h compilation it would probably be **heavy** that is always helpful.

**Comment by laurmaedje** [comment]
If the unformatted document compiles very fast, it's also possible that paragraph layout becomes super slow due to a particular quirk in the paragraph implementation that's been fixed recently. I would be curious if you could try to compile with the main branch.

**Comment by chrarnold** [comment]
I will try a few things and let you guys know yes thanks for the input so far!

To be clear: I am formatting around 50% of the 57 million characters in the document with the same (!) Function: a simple underline with a few custom parameters. Currently, I re-define it every time, which may also contribute to the slowness.

**Comment by Enivex** [comment]
> I will try a few things and let you guys know yes thanks for the input so far!
> 
> To be clear: I am formatting around 50% of the 57 million characters in the document with the same (!) Function: a simple underline with a few custom parameters. Currently, I re-define it every time, which may also contribute to the slowness.

What do you mean by redefining it?

**Comment by Enivex** [comment]
I don't think this is particularly actionable, and OP hasn't replied in 2 years
